### PR TITLE
 Throw warning instead of error for non-constant values in check-model stages

### LINKED ISSUE #3720: --check-models Fatal failure at CVC4/src/theory/theory_engine.cpp:2233
Hi,
For this formula,
```
(assert (_ emp Int Int))
(check-sat)
```
cvc4 `--check-models` throws out a fatal failure:
```
Fatal failure within void CVC4::TheoryEngine::checkTheoryAssertionsWithModel(bool) at CVC4/src/theory/theory_engine.cpp:2233
Internal error detectedTHEORY_SEP has an asserted fact that the model doesn't satisfy.
The fact: (emp _emp1 _emp2)
Model value: (emp 0 0)
Aborted
```

OS: Ubuntu 18.04
Revision: ea43556

### LINKED ISSUE #3729: --check-models Fatal failure at CVC4/src/smt/smt_engine.cpp:4738
Hi,
For this formula,
```
(declare-fun a () Real)
(assert (= a (sin 1.0)))
(check-sat)
```
`cvc4 --check-models` throws out a fatal failure:
```
Fatal failure within void CVC4::SmtEngine::checkModel(bool) at CVC4/src/smt/smt_engine.cpp:4738
Internal error detectedSmtEngine::checkModel(): ERRORS SATISFYING ASSERTIONS WITH MODEL:
model value for a
             is (sin 1)
and that is not a constant (.isConst() == false).
Run with `--check-models -v' for additional diagnostics.
Aborted
```
with `--check-models -v' :
```
minisat: Incremental solving is forced on (to avoid variable elimination) unless using internal decision strategy.
============================[ Search Statistics ]==============================
| Conflicts |          ORIGINAL         |          LEARNT          | Progress |
|           |    Vars  Clauses Literals |    Limit  Clauses Lit/Cl |          |
===============================================================================
===============================================================================
SmtEngine::checkModel(): generating model
(model
Fatal failure within CVC4::Expr CVC4::SmtEngine::getValue(const CVC4::Expr&) const at CVC4/src/smt/smt_engine.cpp:4283
Check failure
 m->hasApproximations() || resultNode.getKind() == kind::LAMBDA || resultNode.isConst()
Aborted
```

OS: Ubuntu 18.04
Revision: 618282e

### PR DESCRIPTION
**Pull request by ajreynol**
Fixes #3729 and fixes #3720. 

This updates two more stages of check-model (checking whether values assigned to terms are constants and internally checking whether assertions belonging to theories) to only throw warnings when a term/assertion has a non-constant value in the model.  This is to accomodate cases where check-model is infeasible.

### CODE REVIEW

File: src/smt/smt_engine.cpp [hunk 1]
```diff
@@ -4785,16 +4785,19 @@ void SmtEngine::checkModel(bool hardFailure) {
...
-               "MODEL:"
-            << endl
-            << "model value for " << func << endl
-            << "             is " << val << endl
-            << "and that is not a constant (.isConst() == false)." << endl
-            << "Run with `--check-models -v' for additional diagnostics.";
+      else if (!val.isConst())
+      {
+        // This is only a warning since it could have been assigned an
+        // unevaluable term (e.g. an application of a trancendental function).
```
Comment (4tXJ7f) [inline src/smt/smt_engine.cpp]:
```suggestion
        // unevaluable term (e.g. an application of a transcendental function).
```

File: src/theory/theory_engine.cpp [hunk 2]
```diff
@@ -2268,14 +2268,26 @@ void TheoryEngine::checkTheoryAssertionsWithModel(bool hardFailure) {
...
-                << "Model value: " << val << endl;
+            if (val == d_false)
+            {
+              // Always an error if it is false
+              InternalError() << ss.str();
+            }
+            else
+            {
+              // Otherwise just a warning. Notice this case may happen for
+              // assertions with unevaluable operators, e.g. trancendental
```
Comment (4tXJ7f) [inline src/theory/theory_engine.cpp]:
```suggestion
              // assertions with unevaluable operators, e.g. transcendental
```

File: src/theory/theory_engine.cpp [hunk 3]
```diff
@@ -2268,14 +2268,26 @@ void TheoryEngine::checkTheoryAssertionsWithModel(bool hardFailure) {
         Node val = getModel()->getValue(assertion);
         if (val != d_true)
         {
+          std::stringstream ss;
+          ss << theoryId
+             << " has an asserted fact that the model doesn't satisfy." << endl
+             << "The fact: " << assertion << endl
+             << "Model value: " << val << endl;
           if (hardFailure)
```
Comment (4tXJ7f) [inline src/theory/theory_engine.cpp]: Why do we only check this when `hardFailure` is enabled?

Comment (ajreynol) [reply src/theory/theory_engine.cpp]: I think hardFailure is impacted by whether the user asked us to check model explicitly or whether this is an internal check.

**Pull request merged**